[interp] Add intrinsics for some common operations

BrzVlad opened the pull request:
These are mostly around string operations and are meant to be a quick boost in performance for some common blazor scenarios, around rendering and json parsing, before interp optimizations catch up. Even if the interpreter will have more advanced optimization passes, it is likely that it would still be much slower on these low level operations. Long term we would have methods like these AOT-ed, but, depending on how easy to maintain and how effective these turn out to be, we could also stick to intrinsifying some bcl code in the future.

BrzVlad committed:
[interp] Add intrinsic for Marvin.Block

BrzVlad committed:
[interp] Intrinsify ConvertAllAsciiCharsInUInt32ToUppercase

BrzVlad committed:
[interp] Intrinisfy UInt32OrdinalIgnoreCaseAscii

BrzVlad committed:
[interp] Intrinsify UInt64OrdinalIgnoreCaseAscii

BrzVlad committed:
[interp] Add intrinsic for UInt32ToDecStr

BrzVlad committed:
[interp] Add intrinsic opcode for WidenAsciiToUtf16

EgorBo commented on src/mono/mono/mini/interp/intrinsics.c:
```diff
@@ -0,0 +1,140 @@
+	guint32 div = a / b;
+	*result = a - (div * b);
+	return div;
+}
```
can be just
```c
*result = a % b;
return a / b;
```
C/C++ compilers are able to recognize a single div there https://godbolt.org/z/-7jsRb

EgorBo commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -1479,6 +1479,37 @@ interp_handle_intrinsics (TransformData *td, MonoMethod *target_method, MonoClas
 		*op = MINT_INTRINS_BYREFERENCE_GET_VALUE;
+	} else if (in_corlib && !strcmp (klass_name_space, "System") && !strcmp (klass_name, "Marvin")) {
+		if (!strcmp (tm, "Block"))
+			*op = MINT_INTRINS_MARVIN_BLOCK;
```
I'd just intrinsify `String.GetHashCode` and implement it using the non-random algorithm like this: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L781-L783 should be way faster and that type of attack usually back-end only

BrzVlad replied:
Doesn't sound like a good idea to have the interpreter just use a different hash function compared to jit/aot or coreclr. Arguably, maybe I should intrinsify both `Marvin.ComputeHash32` and  `Marvin.ComputeHash32OrdinalIgnoreCase`.

EgorBo replied:
I'd suggest to use this one for all Xamarin workloads and all modes including AOT.

also it doesn't matter if two different implementations return different hash codes - string.GetHashCode already returns different hash codes for the same input between sessions - it uses a random seed.

akoeplinger replied:
@EgorBo I think we should discuss that in a separate issue.

MichalStrehovsky commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -1479,6 +1479,37 @@ interp_handle_intrinsics (TransformData *td, MonoMethod *target_method, MonoClas
+	} else if (in_corlib && !strcmp (klass_name_space, "System") && !strcmp (klass_name, "Marvin")) {
+		if (!strcmp (tm, "Block"))
+			*op = MINT_INTRINS_MARVIN_BLOCK;
+	} else if (in_corlib && !strcmp (klass_name_space, "System.Text.Unicode") && !strcmp (klass_name, "Utf16Utility")) {
```
Is this going to silently regress perf if Utf16Utility is ever refactored and things in it shuffled/renamed? This is not a public API so the risk of that is pretty high.

Cc @GrabYourPitchforks

BrzVlad replied:
Unfortunately yes. We should spot potential regressions with the daily perf runs that we will do. Hopefully we will not make use of this technique on a large enough scale for it to become problematic.

akoeplinger replied:
can we mark the method with `[Intrinsic]` or add some comment to make the chance of this happening a bit less likely?

BrzVlad committed:
[interp] Rename intrinsics to interp-intrins

There seem to be some build failures on windows. Probably because of conflict with intrinsics source files from mini.

EgorBo commented on src/libraries/System.Private.CoreLib/src/System/Marvin.cs:
```diff
@@ -215,6 +215,7 @@ public static int ComputeHash32(ref byte data, uint count, uint p0, uint p1)
         }
 
+        [Intrinsic]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
in theory this attribute slows ryujit a bit for no reason, I'd just leave a comment or wrap it with #if mono

BrzVlad committed:
[interp] Add comments

To prevent bcl and interp intrinsics going out of sync.
